ResizeListenerService & BreakpointService: Allow user defined breakpoints

ScarletKuro opened the pull request:
## Description
Fixes this: https://github.com/MudBlazor/MudBlazor/pull/7065#issuecomment-1604603646

We made a slight breaking change in the public API of `ResizeOptions` by switching from `Dictionary<string, int>` to `Dictionary<Breakpoint, int>`. This change is an improvement over the previous broken behavior. In the previous implementation, setting values would have no effect on the `BreakpointService` and `ResizeListenerService` except when you use `IBreakpointService.SubscribeAsync(Action<Breakpoint> callback, ResizeOptions? options)` overload..

The code now handles this change by always cloning the `ResizeOptions` to ensure that we don't mutate the original options injected via dependency injection (`services.Configure(Action<ResizeOptions>...)`). Additionally, we need now a proper `GetHashCode` method for `ResizeOptions` to ensure the correctness of our tests. This is important because we use indexers where the key is a `ResizeOption`, and since we clone `ResizeOptions`, the key would not be found otherwise.

Lastly, the `ResizeListenerService` is marked as `[Obsolete]`. According to this PR https://github.com/MudBlazor/MudBlazor/pull/2658 the `ResizeListenerService` was suppressed by `BreakpointService`.

## How Has This Been Tested?
Existing tests, debugging.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.

ScarletKuro committed:
ResizeListenerService & BreakpointService: Allow user defined breakpoints

ScarletKuro committed:
Remove empty line

ScarletKuro commented:
Don't want to look like a guy who comes and rewrites everything what he sees, but this is also a good candidate for a service rework.
The PR https://github.com/MudBlazor/MudBlazor/pull/2658 says:
> This PR split the existing functionality into two distinct services `IResizeListenerService` and `IBreakpointService`. While the first one returns an object with the window size, the second one provides the mechanism for the breakpoints.

However, I fail to understand why this cannot be consolidated into a single service. The code is identical, and both services perform the same functions, with the only difference being that one subscribes to `BrowserWindowSize` while the other subscribes to `Breakpoint`, as mentioned above.

Also, tjscience comment:
> Anyone familiar with ResizeService? Why was SubscribeAsync and UnsubscribeAsync written with Action<BrowserWindowSize> instead of Func<BrowserWindowSize, Task>? This prevents the ability to (easily) have async in the callback.

I wholeheartedly agree with this. While it is possible to achieve it with a workaround, but then whole service would be full of workarounds which is not ok to me.

Also I took hours of searching, asking and debugging to understand why we have almost 3 identical services which is bad for newcomers. Which is why the `ResizeListenerService` should be removed and the left two merged in two with proper documentation.

ScarletKuro committed:
FIx comment

ScarletKuro commented:
Worth to mention that `ResizeOptions` has 3 options as `EnableLogging`, `SuppressInitEvent`, `NotifyOnBreakpointOnly` yet none of this is used (except last one in **old** service) which shows how abandoned the current state all this has.

ScarletKuro commented:
Visually no regression on `components/hidden#listening-to-browser-window-resize-events` and `https://localhost:5001/components/breakpointprovider#api`

ScarletKuro commented:
> Worth to mention that `ResizeOptions` has 3 options as `EnableLogging`, `SuppressInitEvent`, `NotifyOnBreakpointOnly` yet none of this is used (except last one in **old** service) which shows how abandoned the current state all this has.

Apparently, it is used when passed to the JavaScript side. To be honest, this should be documented, as the IDE indicates that there are no read references, giving a false sense of safety that it can be freely removed. Additionally, it should be protected from trimming in such cases. I am unsure whether we should address this in this pull request or separately.

henon commented:
> Don't want to look like a guy who comes and rewrites everything what he sees, but this is also a good candidate for a service rework. [...]

I really don't mind. We all know that much of the code base needs rewriting. We threw this together in a haste at a time where we didn't yet fully understand best practices in Blazor. Also, we didn't even dream of the success the library now has which adds extra gravity to every design decision. Initially we really built this library simply to fulfill needs we had and most contributors did the same. 

If I had more time on my hands I'd also love to rewrite/redesign some things. In any case I love the fact that you have that time and energy at the moment. 

> I am unsure whether we should address this in this pull request or separately.

If you want to do it in this PR I don't mind at all. Your decision.

henon commented on src/MudBlazor.UnitTests/Mocks/MockResizeListenerService.cs:
```diff
@@ -5,6 +5,7 @@
 namespace MudBlazor.UnitTests.Mocks
 {
 #pragma warning disable CS1998 // Justification - Implementing IResizeListenerService
+    [Obsolete("Suppressed by BreakpointService")]
```
Did you mean Replaced instead of Suppressed?

henon replied:
Or Superceded

ScarletKuro replied:
> Or Superceded

Yea.. I will replace on "replaced" tho, much easier word to understand.

henon commented on src/MudBlazor/Services/ResizeListener/BreakpointService .cs:
```diff
@@ -148,10 +139,19 @@ public async Task<BreakpointServiceSubscribeResult> SubscribeAsync(Action<Breakp
 
             options ??= _options;
-            if (options.BreakpointDefinitions == null || options.BreakpointDefinitions.Count == 0)
+
+            //Always wrap in new instance of ResizeOptions and clone values no matter what
+            //Because the options can come from the _options(IOptions<ResizeOptions>) - these are the options that user sets when adding BreakpointService in DI
+            //Only user is allowed to modify these settings, service should not touch that reference and change it or we get nasty bugs
+            //If we don't always wrap then we would need to write tedious code like if null then wrap, then check the references of options and _options are same and make logic when we can't modify and when we can etc.
```
Awesome, I like such comments helping the reader of the code understand better.

Technically this is not `wrapping` it is `copying` or `cloning`, if I am not mistaken. Wrapping would be if we create a new instance that holds the old instance internally.

ScarletKuro replied:
Yeah, it's cloning, I will adjust the comment

ScarletKuro replied:
Maybe it would be even better to use record here, that already has the cloning logic, equality and gethascode etc, but I think I will consider this when I will rewrite the service.

ScarletKuro committed:
Add Clone extension, update comments

ScarletKuro commented:
I have added an extension method called Clone for `ResizeOptions`. What are your thoughts? I think it's worth even thought this is service is probably the only one who is going to use it. Actually I'm currently doing a WIP of new service and I plan to use it too, and since we keep old for compatibility it's two now :) It should be trimmed anyway if the services won't be used.

Updated the comments too.

henon commented:
Bugs hate you @ScarletKuro because you methodically exterminate them all :D
